python-semantic-release/python-semantic-release
# fix: path traversal prevention auto-failure on windows

## Purpose


- Resolves: #994
- Ensures all testing is cross-platform to include windows
- Adds automated testing of windows compatibility to PR & Main workflows to prevent an issue like 994 from occurring again simply because of lack of testing on a supported platform

## Rationale


See #1014, and I did have to adjust the testing to accomidate windows for some weird compatibility bugs that caused tests to fail.  (1) GitPython under the hood uses `getpass.getuser()` which will fail on windows unless the `USERNAME` variable is set in the environment.  (2) The use of `NamedTemporaryFile()` for testing `netrc` was not compatible with Windows because it restricts the file to one reader at a time and we were leaving the file open (a second read is allowed on linux).  (3) Some expected filepaths were hardcoded strings that represented linux only which were passed through `Path()` to auto-convert path separators.

Lastly, since the testing on windows takes around 10-15 minutes to complete, it was infeasible to integrate windows testing directly into the matrix testing as part of the PR.  Instead, I have it testing the oldest version of python on windows and then matrix testing across all the versions before release.  Trying to get the best of both worlds with a little bit of increased rigor.

## How did you test?

I ran a bunch of pipelines on the windows platform. Then went through each error of the tests to evaluate why it was failing. Ultimately, I did not add any new tests but I did ensure that all test cases run on both windows and maintained compatibility on linux.  This is demonstrated by the pipeline below.

## How to Verify


Review the pipeline results below since we test both windows and linux now in the PR pipeline.  This can still be replicated if you check out this pr and then run `pytest` locally.
